File: BrainfuckInterpreter.js
/**
 * BrainfuckInterpreter.js
 *
 * Interprets Brainfuck code.
 *
 * example usage:
 *
 * const bf = new BfInterpreter();
 * console.log(bf.interpret('++++++[>++++++++++<-]>+++++.'));
 *
 * @author Maxamilian Demian <max@maxdemian.com>
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/BrainfuckInterpreter
 */

class BfInterpreter {
	#updateBlockValue(increase = true) {
		if (increase) {
			if (this.blocks[this.pointer] === 255) {
				this.blocks[this.pointer] = 0;
			} else {
				this.blocks[this.pointer]++;
			}
		} else {
			if (this.blocks[this.pointer] === 0) {
				this.blocks[this.pointer] = 255;
			} else {
				this.blocks[this.pointer]--;
			}
		}
	}

	#updatePointer(increase = true) {
		if (increase) {
			this.pointer++;

			if (typeof this.blocks[this.pointer] === 'undefined') {
				this.blocks[this.pointer] = 0;
			}
		} else if (this.pointer > 0) {
			this.pointer--;
		}
	}

	#getPositionOfClosingBracket(chars) {
		let openBracketCount = 0;

		for (let i = 0; i < chars.length; i++) {
			if (chars[i] === '[') {
				openBracketCount++;
			} else if (chars[i] === ']') {
				if (openBracketCount > 0) {
					openBracketCount--;
				} else {
					return i;
				}
			}
		}
	}

	#handleLoop(chars) {
		const closingBracketPos = this.#getPositionOfClosingBracket(chars);
		const loopCodeChars = chars.slice(0, closingBracketPos);

		// the recursive call to this.#handleInterpret can change the value of this.pointer, so we need to put it in a new var first
		const loopPointer = this.pointer;

		while (this.blocks[loopPointer] !== 0) {
			// we spread the loopCodeChars into a new array to get a new instance instead of passing by reference
			this.#handleInterpret([...loopCodeChars]);
		}

		return chars.slice(closingBracketPos + 1);
	}

	#handleOutput() {
		this.output += String.fromCharCode(this.blocks[this.pointer]);
	}

	#handleInput() {
		this.blocks[this.pointer] = this.input.charCodeAt(this.inputPointer);

		this.inputPointer++;
	}

	#handleInterpret(chars) {
		while (chars.length) {
			let command = chars.shift();

			switch(command) {
				case '+':
					this.#updateBlockValue();

					break;
				case '-':
					this.#updateBlockValue(false);

					break;
				case '>':
					this.#updatePointer();

					break;
				case '<':
					this.#updatePointer(false);

					break;
				case '[':
					chars = this.#handleLoop(chars);

					break;
				case '.':
					this.#handleOutput();

					break;
				case ',':
					this.#handleInput();

					break;
				default:
					// any other character is considered a comment
			}
		}

		return this.output;
	}

	interpret(code = '', input = '') {
		this.input = input;
		this.inputPointer = 0;
		this.blocks = [0];
		this.pointer = 0;
		this.output = '';

		return this.#handleInterpret(code.split(''));
	}
}

File: BrainfuckInterpreter.pl
#!/usr/bin/perl
#
# BrainfuckInterpreter.pl
#
# Interprets Brainfuck code.
#
# example usage
#
# my $bf = BfInterpreter->new;
# print $bf->interpret('++++++[>++++++++++<-]>+++++.');
#
# Author: Maxamilian Demian
#
# https://www.maxodev.org
# https://github.com/Maxoplata/BrainfuckInterpreter

use strict;
use warnings;

# package definition
{
	package BfInterpreter;

	sub new {
		my ($class, $args) = @_;

		my $self = bless {
			input => '',
			inputPointer => 0,
			blocks => [0],
			pointer => 0,
			output => '',
		}, $class;
	}

	sub _updateBlockValue {
		my ($self, $increase) = @_;

		if ($increase == 1) {
			if ($self->{blocks}[$self->{pointer}] == 255) {
				$self->{blocks}[$self->{pointer}] = 0;
			} else {
				$self->{blocks}[$self->{pointer}]++;
			}
		} else {
			if ($self->{blocks}[$self->{pointer}] == 0) {
				$self->{blocks}[$self->{pointer}] = 255;
			} else {
				$self->{blocks}[$self->{pointer}]--;
			}
		}
	}

	sub _updatePointer {
		my ($self, $increase) = @_;

		if ($increase == 1) {
			$self->{pointer}++;

			if (!defined $self->{blocks}[$self->{pointer}]) {
				$self->{blocks}[$self->{pointer}] = 0;
			}
		} elsif ($self->{pointer} > 0) {
			$self->{pointer}--;
		}
	}

	sub _getPositionOfClosingBracket {
		my ($self, @chars) = @_;

		my $openBracketCount = 0;

		for (my $i = 0; $i < scalar(@chars); $i++) {
			if ($chars[$i] eq '[') {
				$openBracketCount++;
			} elsif ($chars[$i] eq ']') {
				if ($openBracketCount > 0) {
					$openBracketCount--;
				} else {
					return $i;
				}
			}
		}
	}

	sub _handleLoop {
		my ($self, @chars) = @_;

		my $closingBracketPos = $self->_getPositionOfClosingBracket(@chars);
		my @loopCodeChars = @chars[0..($closingBracketPos - 1)];

		# the recursive call to $self->handleInterpret can change the value of $self->{pointer}, so we need to put it in a new var first
		my $loopPointer = $self->{pointer};

		while ($self->{blocks}[$loopPointer] != 0) {
			$self->handleInterpret(@loopCodeChars);
		}

		return @chars[($closingBracketPos + 1)..$#chars];
	}

	sub _handleOutput {
		my ($self) = @_;

		$self->{output} .= chr($self->{blocks}[$self->{pointer}]);
	}

	sub _handleInput {
		my ($self) = @_;

		$self->{blocks}[$self->{pointer}] = ord(substr($self->{input}, $self->{inputPointer}, 1));

		$self->{inputPointer}++;
	}

	sub handleInterpret {
		my ($self, @chars) = @_;

		while (scalar(@chars) > 0) {
			my $command = shift @chars;

			if ($command eq '+') {
				$self->_updateBlockValue(1);
			} elsif ($command eq '-') {
				$self->_updateBlockValue(0);
			} elsif ($command eq '>') {
				$self->_updatePointer(1);
			} elsif ($command eq '<') {
				$self->_updatePointer(0);
			} elsif ($command eq '[') {
				@chars = $self->_handleLoop(@chars);
			} elsif ($command eq '.') {
				$self->_handleOutput();
			} elsif ($command eq ',') {
				$self->_handleInput();
			}

			# any other character is considered a comment
		}

		return $self->{output};
	}

	sub interpret {
		my ($self, $code, $input) = @_;

		$code ||= '';
		$input ||= '';

		$self->{input} = $input;
		$self->{inputPointer} = 0;
		$self->{blocks} = [0];
		$self->{pointer} = 0;
		$self->{output} = '';

		return $self->handleInterpret(split('', $code));
	}
}
